File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
class Solution {
    int answer = 0;
        
    public int solution(int[] numbers, int target) {
        
        dfs(numbers, target, 0, 0);
        
        
        return answer;
    }
    
    private void dfs(int[] numbers, int target, int index, int sum){
        if(index == numbers.length){ // 배열을 다 돌았을때
            if(sum == target){ // 값이 타겟팅된다면 정답++
                answer++;
            }
            return; // 값이 맞던 안맞던 배열을 다 돌면 종료해야 하므로 else 없이 리턴
                
            
        }
        dfs(numbers, target, index + 1, sum + numbers[index]); //더하거나
        dfs(numbers, target, index + 1, sum - numbers[index]);// 빼거나 
        
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
import sys
sys.setrecursionlimit(10**6)

def solution(numbers, target):
    answer = 0
    n = len(numbers) # 길이를 미리 저장
    
    # 함수 안에 함수를 만들면 외부 변수(numbers, target)를 그냥 쓸 수 있음
    def dfs(idx, current_sum):
        nonlocal answer # solution의 answer 변수를 갖다 쓰겠다는 뜻
        
        # 1. 탈출 조건: 끝까지 다 탐색했을 때
        if idx == n:
            if current_sum == target:
                answer += 1
            return # 여기서만 리턴해야 함!
        
        # 2. 수행 동작: 더하거나 빼거나 (재귀)
        dfs(idx + 1, current_sum + numbers[idx])
        dfs(idx + 1, current_sum - numbers[idx])

    # DFS 시작 (인덱스 0, 합계 0)
    dfs(0, 0)
    
    return answer